Add user profile patch route

AlexisDougherty13 opened the pull request:
Added a patch route for editing user profile. Was unable to test to be sure it works but I think it will. Definitely compiles. I'll clean it up a bit when I'm able to test it for sure.

AlexisDougherty13 committed:
Add user profile patch route

AlexisDougherty13 committed:
Fixed firstname glitch issue

chris-brugal reviewed (approved):
looks good

WillBAnders commented on backend/src/router.go:
```diff
@@ -349,3 +351,48 @@ func getProfile(c *gin.Context) {
+	
+	var edits ProfileUpdateData
+	if err := c.ShouldBindJSON(&edits); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
```
Minor nitpick - for consistency, let's stick to directly using `400` as the status rather than the constant. We can open an issue to make this change to all of the routes later; didn't realize this was a thing.

AlexisDougherty13 replied:
Fixed to a 400 and added the documentation. I had planned on adding it soon, just wanted to get the init stuff working before I made docs and then would need to make major changes to it.

WillBAnders commented on backend/src/router.go:
```diff
@@ -349,3 +351,48 @@ func getProfile(c *gin.Context) {
+		return
+	}
+	
+	var users []User
```
Move declaration to directly before the `Find` call.

AlexisDougherty13 replied:
Done :-)

WillBAnders commented on backend/src/router.go:
```diff
@@ -349,3 +351,48 @@ func getProfile(c *gin.Context) {
+	}
+	
+	DB.Find(&users, "username = ?", claims.Username)
+	DB.Model(&users[0]).Updates(edits)
```
I saw the latest commit regarding some issue with Gorm here and took a closer look. [According to the docs](https://gorm.io/docs/update.html#Updates-multiple-columns), `Updates` with structs will not change default values which is what we need (that way a user can remove an email or something). We also need to be careful about distinguishing between default values passed in the body to be updated versus default values applied from binding. Not sure the best way to handle this; it's probably a separate issue.

WillBAnders commented on backend/src/router.go:
```diff
@@ -349,3 +351,48 @@ func getProfile(c *gin.Context) {
+	Phone     string `json:"phone"`
+}
+
+// ADD DESCRIPTION
```
The route documentation should be done in this commit, that way Chris knows the intended schema for the route so he can use it on the frontend.

AlexisDougherty13 replied:
Done, although it's in progress due to adding tutors

WillBAnders reviewed (changes requested):
Only change I'm going to request is adding in the documentation for the route; it should be done here rather than in a separate PR. The other styling changes would be nice too.

The handling of update seems to be a bit of a problem, you're welcome to tackle that here or we can make an issue for it an handle it in a later PR - just make sure to note this limitation in the doc comment.

AlexisDougherty13 committed:
Add changes per review recommendations; Close #96

chris-brugal committed:
Feature/profile (#99)

* refactor: updated old prof page to tutor page and created real profile page

* feat: added edit profile page

* Delete TutorPage.js

We changed the file names. TutorPage is now CoursePage and what used to be CoursePage is CoursesPage.

* refactor: readded tutorpage

Co-authored-by: cole-kitroser <70605063+cole-kitroser@users.noreply.github.com>

AlexisDougherty13 committed:
Add bio and availablility editting for /profile; Close #77

WillBAnders commented:
Rebase and should be good to merge. We'll handle backend tests for it as part of finalizing the specification for default values (see https://github.com/WillBAnders/TutorsVILLE/pull/97#discussion_r837852527).

AlexisDougherty13 committed:
Add add/remove for tutor's classes; Close #78

WillBAnders commented:
Split off later commits for other frontend work into a separate PR. The route implementation still needs some work but we can address the edge cases in combination with implementing tests.
